fix: improvements to `gptme-util chats read/search`

ErikBjare opened the pull request:
<!-- ELLIPSIS_HIDDEN -->


> [!IMPORTANT]
> Enhance `chats` functionality in CLI and `chats.py` with context and match limit options, improving search and read capabilities.
> 
>   - **Behavior**:
>     - `list_chats()` in `chats.py` now displays message count when metadata is not included.
>     - `search_chats()` in `chats.py` supports `context_lines` and `max_matches` for better match display.
>     - `read_chat()` in `chats.py` adds `context_messages` and `start_message` options for message filtering.
>   - **CLI**:
>     - `chats_search` and `chats_read` commands in `cli.py` now accept `--context` and `--matches` options for enhanced search and read functionalities.
>     - `chats_read` command in `cli.py` supports `--start` option to specify starting message.
> 
> <sup>This description was created by </sup>[<img alt="Ellipsis" src="https://img.shields.io/badge/Ellipsis-blue?color=175173">](https://www.ellipsis.dev?ref=ErikBjare%2Fgptme&utm_source=github&utm_medium=referral)<sup> for 8df3fde1612d5051935efbdf27f13a14f09dfd22. It will automatically update as commits are pushed.</sup>


<!-- ELLIPSIS_HIDDEN -->

TimeToBuildBob commented:
The test failure is fixed in PR #2152 — updated the test assertions to match the actual output format (`(1 msgs)` vs `Messages: 1`). Once that merges, CI should be green.

ErikBjare commented:
@TimeToBuildBob Take over this PR

TimeToBuildBob committed:
fix(chats): improve search/list output with line-based context and message counts

- list_chats: show '(N msgs)' in compact (non-metadata) mode instead of full metadata
- _format_message_with_context: switch from character-based to line-based context
  with line numbers and ANSI highlighting for better code/text readability
- search_chats: improved multi-match display shows message index and role per match;
  rename context_size → context_lines (lines, not chars); update CLI default to 1 line
- tests: update assertions to match new compact list format

Ports the improvements from PR #437 (Erik's original work) onto current master.

TimeToBuildBob committed:
fix(chats): clean up metadata formatting and add tty check for ANSI color

- Replace textwrap.indent(...)[4:] hack with explicit first/rest split
- Guard ANSI highlight codes with sys.stdout.isatty(); fall back to **bold**
  markdown for non-tty outputs (pipes, scripts, redirects)

TimeToBuildBob committed:
fix(chats): address Greptile review findings

- Gate LogManager.load on include_summary only; use conv.messages for
  the simple count display (avoids unnecessary JSONL reads on list calls)
- Use len(query_lower) consistently for end_pos to match how start advances

TimeToBuildBob commented on gptme/tools/chats.py:
```diff
@@ -48,12 +47,15 @@ def list_chats(
+
+        if metadata:
+            print()  # Add a newline between conversations
+            print(f"{i:2}. {textwrap.indent(conv.format(metadata=True), '    ')[4:]}")
```
Fixed in 9e5c817 — replaced `textwrap.indent(...)[4:]` with an explicit split on newlines, printing the first line with the index prefix and remaining lines with consistent indentation.

TimeToBuildBob commented on gptme/tools/chats.py:
```diff
@@ -110,106 +114,154 @@ def search_chats(
+        formatted_lines = []
+        for i, line in enumerate(context_lines_text, start=context_start):
+            if i == match_idx:
+                # Highlight the matching part in the line
```
Fixed in 9e5c817 — `_format_message_with_context` now checks `sys.stdout.isatty()` before emitting ANSI codes; falls back to `**match**` markdown-style highlighting for pipes, scripts, and redirects.

TimeToBuildBob commented on gptme/tools/chats.py:
```diff
@@ -61,12 +60,15 @@ def list_chats(
+            print()  # Add a newline between conversations
+            print(f"{i:2}. {textwrap.indent(conv.format(metadata=True), '    ')[4:]}")
+        else:
+            print(f"{i:2}. {conv.name} ({msg_count} msgs)")
```
Fixed in b4395d8 — `LogManager.load` is now gated on `include_summary`; the simple count display uses `conv.messages` which is already in `ConversationMeta`. No JSONL reads on plain `list` calls.

TimeToBuildBob commented on gptme/tools/chats.py:
```diff
@@ -132,83 +134,85 @@ def search_chats(
+            idx = line_lower.find(query_lower, start)
+            if idx == -1:
+                break
+            line_indices.append((line_idx, idx, idx + len(query)))
```
Fixed in b4395d8 — changed `idx + len(query)` to `idx + len(query_lower)` for consistent use of the lowercased form. For ASCII this is a no-op; for edge-case Unicode (e.g. ß→ss) it's defensively correct.

TimeToBuildBob commented:
Taking over from Erik per his request.

Changes since the last review:
- **Rebased onto master** (was 4 commits behind)
- **Ellipsis: `textwrap.indent(...)[4:]` hack** — replaced with explicit first/rest split (9e5c817)
- **Ellipsis: ANSI color without tty check** — `sys.stdout.isatty()` guard added; falls back to `**bold**` for pipes (9e5c817)
- **Greptile: unnecessary `LogManager.load`** — gated on `include_summary` only; simple count display uses `conv.messages` from `ConversationMeta` (b4395d8)
- **Greptile: `end_pos` inconsistency** — unified to `len(query_lower)` (b4395d8)

CI pending. Will monitor.

TimeToBuildBob commented:
@greptileai review

TimeToBuildBob committed:
fix(chats): merge overlapping context windows; add --metadata to chats list

Address two Greptile review concerns:

- _format_message_with_context: when max_matches > 1 and two matches are
  on nearby lines, their context windows could overlap and duplicate shared
  lines. Now merges adjacent/overlapping windows into a single block, so
  each line appears at most once and both matches are highlighted within it.

- chats list: metadata parameter was previously always False from the CLI
  (no flag to set it). Add --metadata flag so users can still see IDs,
  model, cost, and token counts without reaching for --json.

TimeToBuildBob commented:
Addressed the two Greptile 4/5 concerns (commit b5762395):

**1. Duplicate context blocks (`_format_message_with_context`)** — when `max_matches > 1` and two matches fell in overlapping context windows, the shared lines were printed twice. Fixed by merging adjacent/overlapping windows into a single block before rendering, so each source line appears at most once with all matches highlighted within it.

**2. `chats list` hides IDs by default** — the `metadata` parameter was now correctly respected but the CLI had no flag to set it. Added `--metadata` flag to `gptme-util chats list` so users can access IDs, model, cost, and token counts without reaching for `--json`.

TimeToBuildBob commented:
@greptileai review

TimeToBuildBob commented:
Taking this over — will review the current state, rebase/merge master, and push the updated branch.

TimeToBuildBob committed:
fix(test): mock sys.stdout.isatty for ANSI highlight tests

Tests asserting ANSI escape codes fail in CI (non-tty) because
_highlight_line falls back to **bold** markdown when stdout is not
a tty. Use monkeypatch to force isatty=True in the two affected tests.

TimeToBuildBob commented:
Fixed the CI failure (commit 0aae76d):

`test_format_message_basic` and `test_format_message_case_insensitive_highlight` were asserting ANSI escape codes (`\033[1;31m`), but `_highlight_line` falls back to `**bold**` markdown when `sys.stdout.isatty()` returns False — which it does in CI. Added `monkeypatch.setattr(sys.stdout, "isatty", lambda: True)` to both tests so they exercise the ANSI code path explicitly.

TimeToBuildBob committed:
test(chats): cover read and search cli options

TimeToBuildBob commented:
@greptileai review

TimeToBuildBob commented on tests/test_util_cli.py:
```diff
@@ -123,10 +123,8 @@ def test_chats_list(tmp_path, mocker):
     assert "Chat One" in result.output
     assert "Chat Two" in result.output
-    assert "2024-01-01-chat-one" in result.output
-    assert "2024-01-01-chat-two" in result.output
-    assert "Messages: 1" in result.output  # First chat has 1 message
-    assert "Messages: 2" in result.output  # Second chat has 2 messages
+    assert "(1 msgs)" in result.output  # First chat has 1 message
+    assert "(2 msgs)" in result.output  # Second chat has 2 messages
     assert "Using browser tool with" not in result.output
```
Added coverage in faea7f8: `chats search --matches 2 --context 1` now asserts both match blocks and line context, and `chats read --start 2 --context 1 --limit 2` now verifies the preceding message is included.
